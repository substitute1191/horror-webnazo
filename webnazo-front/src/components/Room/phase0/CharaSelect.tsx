import { useState } from "react"
import useCharaSelect from "./useCharaSelect"
import TeamNameFormModal from "./TeamNameFormModal"

const CharaSelect = () => {
  const { myChara, otherChara, handleChange, isAllSelected } = useCharaSelect()
  const [isShowModal, setIsShowModal] = useState(false)

  const closeModal = () => setIsShowModal(false)

  return (
    <div>
      <TeamNameFormModal isOpen={isShowModal} onRequestClose={closeModal} />
      <div className="bg-home-bg object-fit flex h-screen flex-col items-center justify-center bg-cover text-6xl">
        <form id="playerSelect">
          <div className="mb-8">
            <input
              type="radio"
              name="player"
              id="player1"
              className="peer/player1 hidden"
              onChange={() => handleChange(1)}
              checked={myChara === 1}
              disabled={otherChara === 1 ? true : false}
            />
            <label
              htmlFor="player1"
              className="font-onryou block rounded p-4 text-white peer-checked/player1:bg-slate-500 peer-disabled/player1:cursor-not-allowed peer-disabled/player1:opacity-50"
            >
              プレイヤー１
            </label>
          </div>
          <div className="mb-8">
            <input
              type="radio"
              name="player"
              id="player2"
              className="peer/player2 hidden"
              onChange={() => handleChange(2)}
              checked={myChara === 2}
              disabled={otherChara === 2 ? true : false}
            />
            <label
              htmlFor="player2"
              className="font-onryou block rounded p-4 text-white peer-checked/player2:bg-slate-500 peer-disabled/player2:cursor-not-allowed peer-disabled/player2:opacity-50"
            >
              プレイヤー２
            </label>
          </div>
        </form>
        {isAllSelected ? (
          <button className="font-onryou" onClick={() => setIsShowModal(true)}>
            ゲームスタート
          </button>
        ) : null}
      </div>
    </div>
  )
}

export default CharaSelect
