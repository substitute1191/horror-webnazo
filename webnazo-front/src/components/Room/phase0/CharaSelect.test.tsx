/* eslint-disable max-lines */
import CharaSelect from "./CharaSelect"
import { render, screen, waitFor } from "@testing-library/react"
import { userEvent } from "@testing-library/user-event"
import * as jotai from "jotai"
import type * as react from "react"

jest.mock("react-router-dom", () => ({
  useParams: jest.fn((): { roomId: string } => ({ roomId: mockRoomId })),
}))
jest.mock("jotai", () => ({
  ...jest.requireActual<typeof jotai>("jotai"),
  useAtomValue: jest.fn((): string => mockUserId),
}))
jest.mock("react", () => ({
  ...jest.requireActual<typeof react>("react"),
  useContext: jest.fn(() => ({
    socket: { init: null },
    isConnected: false,
  })),
}))
jest.mock("./useCharaSelect", () => ({
  __esModule: true,
  default: () => mockUseCharaSelect(),
}))

const mockRoomId = "mock-room-id"
const mockUserId = "mock-user-id"
const mockUseCharaSelectState = {
  myChara: 0,
  otherChara: 0,
  handleChange: jest.fn(),
  isAllSelected: false,
}
const mockUseCharaSelect = jest.fn(() => mockUseCharaSelectState)

describe("CharaSelect単体のテスト", () => {
  const user = userEvent.setup()
  beforeEach(() => {
    jest.clearAllMocks()
    mockUseCharaSelectState.myChara = 0
    mockUseCharaSelectState.otherChara = 0
    mockUseCharaSelectState.isAllSelected = false
  })

  it("ボタンのレンダリング確認", () => {
    render(<CharaSelect />)

    const player1 = screen.getByText("プレイヤー１")
    const player2 = screen.getByText("プレイヤー２")
    const hiddenStartButton = screen.queryByText("ゲームスタート")
    expect(player1).toBeInTheDocument()
    expect(player2).toBeInTheDocument()
    expect(hiddenStartButton).not.toBeInTheDocument()
  })

  it("スタートボタンの出現確認", () => {
    mockUseCharaSelectState.myChara = 1
    mockUseCharaSelectState.otherChara = 2
    mockUseCharaSelectState.isAllSelected = true

    const { container } = render(<CharaSelect />)

    const StartButton = screen.queryByText("ゲームスタート")

    expect(StartButton).toBeInTheDocument()
    expect(container).toMatchSnapshot("両プレイヤー選択後")
  })

  // TODO 一項目の粒度の見直し
  it("プレイヤー選択時の処理確認", async () => {
    const { rerender } = render(<CharaSelect />)
    const player1 = screen.getByLabelText("プレイヤー１")
    const player2 = screen.getByLabelText("プレイヤー２")

    // プレイヤー選択時にhandleChangeが呼ばれることの確認
    await user.click(player1)
    await waitFor(() =>
      expect(mockUseCharaSelectState.handleChange).toHaveBeenCalledWith(1)
    )

    // プレイヤー１を選択した状態でプレイヤー２選択時にhandleChangeが呼ばれることの確認
    mockUseCharaSelectState.myChara = 1
    rerender(<CharaSelect />)
    await user.click(player2)
    await waitFor(() =>
      expect(mockUseCharaSelectState.handleChange).toHaveBeenCalledWith(2)
    )

    // プレイヤー２を選択した状態でプレイヤー２選択時にhandleChangeが呼ばれないことの確認
    mockUseCharaSelectState.myChara = 2
    rerender(<CharaSelect />)
    await user.click(player2)
    await waitFor(() =>
      expect(mockUseCharaSelectState.handleChange).toHaveBeenCalledTimes(2)
    )
  })

  it("相手がプレイヤー１を選択しているときの処理確認", () => {
    mockUseCharaSelectState.otherChara = 1

    const { rerender } = render(<CharaSelect />)

    const player1 = screen.getByLabelText("プレイヤー１")
    const player2 = screen.getByLabelText("プレイヤー２")

    // 相手がプレイヤー１を選択時にこちらがプレイヤー１だけ選択できないことのテスト
    expect(player1).toBeDisabled()
    expect(player2).not.toBeDisabled()

    // 相手がプレイヤー２を選択しているときのテスト
    mockUseCharaSelectState.otherChara = 2
    rerender(<CharaSelect />)

    expect(player1).not.toBeDisabled()
    expect(player2).toBeDisabled()
  })

  it("初期状態のコンポーネントのスナップショットが一致する", () => {
    const { container } = render(<CharaSelect />)
    const StartButton = screen.queryByText("ゲームスタート")

    expect(StartButton).not.toBeInTheDocument()
    expect(container).toMatchSnapshot("初期状態")
  })
})
