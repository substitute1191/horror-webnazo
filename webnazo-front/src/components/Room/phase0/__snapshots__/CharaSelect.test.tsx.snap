// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharaSelect単体のテスト スタートボタンの出現確認: 両プレイヤー選択後 1`] = `
<div>
  <div
    class="bg-home-bg object-fit flex h-screen flex-col items-center justify-center bg-cover text-6xl"
  >
    <form
      id="playerSelect"
    >
      <div
        class="mb-8"
      >
        <input
          checked=""
          class="peer/player1 hidden"
          id="player1"
          name="player"
          type="radio"
        />
        <label
          class="font-onryou block rounded p-4 text-white peer-checked/player1:bg-slate-500 peer-disabled/player1:cursor-not-allowed peer-disabled/player1:opacity-50"
          for="player1"
        >
          プレイヤー１
        </label>
      </div>
      <div
        class="mb-8"
      >
        <input
          class="peer/player2 hidden"
          disabled=""
          id="player2"
          name="player"
          type="radio"
        />
        <label
          class="font-onryou block rounded p-4 text-white peer-checked/player2:bg-slate-500 peer-disabled/player2:cursor-not-allowed peer-disabled/player2:opacity-50"
          for="player2"
        >
          プレイヤー２
        </label>
      </div>
    </form>
    <button
      class="font-onryou"
    >
      ゲームスタート
    </button>
  </div>
</div>
`;

exports[`CharaSelect単体のテスト 初期状態のコンポーネントのスナップショットが一致する: 初期状態 1`] = `
<div>
  <div
    class="bg-home-bg object-fit flex h-screen flex-col items-center justify-center bg-cover text-6xl"
  >
    <form
      id="playerSelect"
    >
      <div
        class="mb-8"
      >
        <input
          class="peer/player1 hidden"
          id="player1"
          name="player"
          type="radio"
        />
        <label
          class="font-onryou block rounded p-4 text-white peer-checked/player1:bg-slate-500 peer-disabled/player1:cursor-not-allowed peer-disabled/player1:opacity-50"
          for="player1"
        >
          プレイヤー１
        </label>
      </div>
      <div
        class="mb-8"
      >
        <input
          class="peer/player2 hidden"
          id="player2"
          name="player"
          type="radio"
        />
        <label
          class="font-onryou block rounded p-4 text-white peer-checked/player2:bg-slate-500 peer-disabled/player2:cursor-not-allowed peer-disabled/player2:opacity-50"
          for="player2"
        >
          プレイヤー２
        </label>
      </div>
    </form>
  </div>
</div>
`;

exports[`Phase0のテスト 「ゲームスタート」出現後のコンポーネントのスナップショットが一致する: 両プレイヤー選択後 1`] = `
<div>
  <div
    class="bg-home-bg object-fit flex h-screen flex-col items-center justify-center bg-cover text-6xl"
  >
    <form
      id="playerSelect"
    >
      <div
        class="mb-8"
      >
        <input
          checked=""
          class="peer/player1 hidden"
          id="player1"
          name="player"
          type="radio"
        />
        <label
          class="font-onryou block rounded p-4 text-white peer-checked/player1:bg-slate-500 peer-disabled/player1:cursor-not-allowed peer-disabled/player1:opacity-50"
          for="player1"
        >
          プレイヤー１
        </label>
      </div>
      <div
        class="mb-8"
      >
        <input
          class="peer/player2 hidden"
          disabled=""
          id="player2"
          name="player"
          type="radio"
        />
        <label
          class="font-onryou block rounded p-4 text-white peer-checked/player2:bg-slate-500 peer-disabled/player2:cursor-not-allowed peer-disabled/player2:opacity-50"
          for="player2"
        >
          プレイヤー２
        </label>
      </div>
    </form>
    <button
      class="font-onryou"
    >
      ゲームスタート
    </button>
  </div>
</div>
`;

exports[`Phase0のテスト 初期状態のコンポーネントのスナップショットが一致する: 初期状態 1`] = `
<div>
  <div
    class="bg-home-bg object-fit flex h-screen flex-col items-center justify-center bg-cover text-6xl"
  >
    <form
      id="playerSelect"
    >
      <div
        class="mb-8"
      >
        <input
          class="peer/player1 hidden"
          id="player1"
          name="player"
          type="radio"
        />
        <label
          class="font-onryou block rounded p-4 text-white peer-checked/player1:bg-slate-500 peer-disabled/player1:cursor-not-allowed peer-disabled/player1:opacity-50"
          for="player1"
        >
          プレイヤー１
        </label>
      </div>
      <div
        class="mb-8"
      >
        <input
          class="peer/player2 hidden"
          id="player2"
          name="player"
          type="radio"
        />
        <label
          class="font-onryou block rounded p-4 text-white peer-checked/player2:bg-slate-500 peer-disabled/player2:cursor-not-allowed peer-disabled/player2:opacity-50"
          for="player2"
        >
          プレイヤー２
        </label>
      </div>
    </form>
  </div>
</div>
`;
